From 336740f72b0789de83dd3373f99380e4d40ef149 Mon Sep 17 00:00:00 2001
From: James Growden <jgrowden@tenstorrent.com>
Date: Fri, 12 Dec 2025 11:17:29 -0500
Subject: [PATCH] drivers: smbus: hook cancel/uncancel to API

Hook the cancel/uncancel interface to the implementation
defined API to perform the cancel/uncancel.

Signed-off-by: James Growden <jgrowden@tenstorrent.com>
---
 drivers/i2c/i2c_ll_stm32_v2.c  |  7 ++++++-
 include/zephyr/drivers/smbus.h | 21 +++++++++++++++++----
 2 files changed, 23 insertions(+), 5 deletions(-)

diff --git a/drivers/i2c/i2c_ll_stm32_v2.c b/drivers/i2c/i2c_ll_stm32_v2.c
index e55a2640791..a62448ee8e5 100644
--- a/drivers/i2c/i2c_ll_stm32_v2.c
+++ b/drivers/i2c/i2c_ll_stm32_v2.c
@@ -1352,6 +1352,11 @@ int i2c_stm32_transaction(const struct device *dev,
 {
 	int ret = 0;
 
+	struct i2c_stm32_data *data = dev->data;
+	if (data->cancelled) {
+		return -EIO;
+	}
+
 #ifdef CONFIG_I2C_STM32_INTERRUPT
 	ret = stm32_i2c_irq_xfer(dev, &msg, next_msg_flags, periph);
 #else
@@ -1367,7 +1372,7 @@ int i2c_stm32_transaction(const struct device *dev,
 	 * which will make the combination of all chunks to look like one big
 	 * transaction on the wire.
 	 */
-	struct i2c_stm32_data *data = dev->data;
+
 	const struct i2c_stm32_config *cfg = dev->config;
 	I2C_TypeDef *i2c = cfg->i2c;
 	const uint32_t i2c_stm32_maxchunk = 255U;
diff --git a/include/zephyr/drivers/smbus.h b/include/zephyr/drivers/smbus.h
index a5b8f2b4040..fe544e80ef7 100644
--- a/include/zephyr/drivers/smbus.h
+++ b/include/zephyr/drivers/smbus.h
@@ -1107,8 +1107,14 @@ __syscall int smbus_cancel(const struct device *dev);
 
 static inline int z_impl_smbus_cancel(const struct device *dev)
 {
-	(void)dev;
-	return 0;
+	const struct smbus_driver_api *api =
+		(const struct smbus_driver_api *)dev->api;
+
+	if (api->smbus_cancel == NULL) {
+		return -ENOSYS;
+	}
+
+	return  api->smbus_cancel(dev);
 }
 
 /**
@@ -1125,10 +1131,17 @@ __syscall int smbus_uncancel(const struct device *dev);
 
 static inline int z_impl_smbus_uncancel(const struct device *dev)
 {
-	(void)dev;
-	return 0;
+	const struct smbus_driver_api *api =
+		(const struct smbus_driver_api *)dev->api;
+
+	if (api->smbus_uncancel == NULL) {
+		return -ENOSYS;
+	}
+
+	return  api->smbus_uncancel(dev);
 }
 
+
 #ifdef __cplusplus
 }
 #endif
-- 
2.34.1

